File: src/Exporter.NX.Translate/DocumentTranslator.cs
// SPDX-License-Identifier: GPL-2.0-only
using System;
using Oblikovati.Exporter.NX.Model;
using Oblikovati.Exporter.NX.Recipe;

namespace Oblikovati.Exporter.NX.Translate
{
    /// <summary>
    /// Top of the translation core: turns an NX-neutral <see cref="NxDocument"/> into a
    /// serializable <see cref="OblikovatiDocument"/>. Assembly translation (.oad)
    /// arrives in M6; until then a non-part document raises a clear error.
    /// </summary>
    public sealed class DocumentTranslator
    {
        public OblikovatiDocument Translate(NxDocument document, ExportReport report)
        {
            if (document == null) throw new ArgumentNullException(nameof(document));
            if (report == null) throw new ArgumentNullException(nameof(report));

            if (document.Kind != NxDocumentKind.Part)
            {
                throw new NotSupportedException(
                    $"document kind '{document.Kind}' is not translatable yet; expected Part");
            }

            return new OblikovatiDocument
            {
                SchemaVersion = 2,
                DocumentType = (int)NxDocumentKind.Part,
                DisplayName = document.DisplayName,
                Model = TranslatePart(document, report),
            };
        }

        private static PartRecipe TranslatePart(NxDocument document, ExportReport report)
        {
            var recipe = new PartRecipe
            {
                Units = new Units
                {
                    Length = document.LengthUnit,
                    Angle = document.AngleUnit,
                },
            };

            foreach (NxExpression expression in document.Expressions)
            {
                recipe.Parameters.Add(ParameterTranslator.Translate(expression));
            }

            // One id space across sketches, points and entities (matches the Go codec).
            var ids = new IdAllocator();
            var sketches = new SketchTranslator(ids, report);
            foreach (NxSketch sketch in document.Sketches)
            {
                int sketchId = ids.Next();
                recipe.Sketches.Add(sketches.Translate(sketch, sketchId));
            }

            var features = new FeatureTranslator(report);
            foreach (NxFeature feature in document.Features)
            {
                FeatureData? translated = features.Translate(feature);
                if (translated != null)
                {
                    recipe.Features.Add(translated);
                }
            }

            return recipe;
        }
    }
}

File: python/oblikovati_exporter_nx/translate/document_translator.py
# SPDX-License-Identifier: GPL-2.0-only
"""Builds a serializable OblikovatiDocument from an NX-neutral NxDocument.

Part documents are self-contained; assembly documents need their occurrences' component
file names from the tree walk (see DocumentExporter), so that path uses
``translate_assembly``.
"""
from __future__ import annotations

from typing import Dict, List

from ..model.document import NxDocument, NxDocumentKind
from ..recipe.document import (
    AssemblyRecipe,
    OblikovatiDocument,
    OccurrenceData,
    PartRecipe,
    Units,
)
from . import parameter_translator, workplane_translator
from .feature_translator import FeatureTranslator
from .id_allocator import IdAllocator
from .report import ExportReport
from .sketch_translator import SketchTranslator


class DocumentTranslator:
    def translate(self, document: NxDocument, report: ExportReport) -> OblikovatiDocument:
        """Translates a part document. Raises for a non-part (use the exporter for assemblies)."""
        if document.kind != NxDocumentKind.PART:
            raise ValueError(
                f"document kind '{document.kind}' is not a part; "
                "assemblies go through DocumentExporter"
            )
        return OblikovatiDocument(
            schema_version=2,
            document_type=int(NxDocumentKind.PART),
            display_name=document.display_name,
            model=self._translate_part(document, report),
        )

    def translate_assembly(
        self, document: NxDocument, occurrences: List[OccurrenceData]
    ) -> OblikovatiDocument:
        """Builds an assembly document from its display info and resolved occurrences."""
        recipe = AssemblyRecipe(
            units=Units(length=document.length_unit, angle=document.angle_unit),
            occurrences=list(occurrences),
        )
        return OblikovatiDocument(
            schema_version=2,
            document_type=int(NxDocumentKind.ASSEMBLY),
            display_name=document.display_name,
            model=recipe,
        )

    @staticmethod
    def _translate_part(document: NxDocument, report: ExportReport) -> PartRecipe:
        recipe = PartRecipe(
            units=Units(length=document.length_unit, angle=document.angle_unit)
        )

        for expression in document.expressions:
            recipe.parameters.append(parameter_translator.translate(expression))

        for plane in document.work_planes:
            recipe.work_features.append(workplane_translator.translate(plane))

        # One id space across sketches, points and entities (matches the Go codec).
        ids = IdAllocator()
        sketches = SketchTranslator(ids, report)
        for sketch in document.sketches:
            sketch_id = ids.next()
            recipe.sketches.append(sketches.translate(sketch, sketch_id))

        # Map each IR feature index to its recipe index so patterns/mirror can remap their
        # source program indices, skipping over any feature that was not translated.
        features = FeatureTranslator(report)
        source_index: Dict[int, int] = {}
        for i, feature in enumerate(document.features):
            translated = features.translate(feature, source_index)
            if translated is not None:
                source_index[i] = len(recipe.features)
                recipe.features.append(translated)

        return recipe
